src/type-check.rkt

```diff
@@ -34,31 +41,47 @@
   (unless (null? errs)
     (raise-herbie-syntax-error "Program has type errors" #:locations errs)))
 
+(define (type-of expr env)
+  (expression->type (datum->syntax #f expr) env
+                    (lambda (stx msg . args)
+                      (error "Unexpected call to error! within type-of"
+                             stx (apply format msg args)))))
+
+
 (define (expression->type stx env error!)
   (match stx
     [(or #`TRUE #`FALSE) 'bool]
     [#`,(? constant? x) 'real]
     [#`,(? variable? x) (dict-ref env x)]
     [#`(,(and (or '+ '- '* '/) op) #,exprs ...)
+     (define t #f)
      (for ([arg exprs] [i (in-naturals)])
        (define actual-type (expression->type arg env error!))
-       (unless (equal? actual-type 'real)
-         (error! stx "~a expects argument ~a of type ~a (not ~a)" op (+ i 1) 'real actual-type)))
-     'real]
+       (if (not t) (set! t actual-type) #f)
+       (unless (equal? t actual-type)
+         (error! stx "~a expects argument ~a of type ~a (not ~a)" op (+ i 1) t actual-type)))
+     ; if we have at least one complex type in exprs we return complex otherwise real
+     t]
     [#`(,(? operator? op) #,exprs ...)
-     (match (get-params op (length exprs))
-       [(list '* each-type)
-        (for ([arg exprs] [i (in-naturals)])
-          (define actual-type (expression->type arg env error!))
-          (unless (equal? actual-type each-type)
-            (error! stx "~a expects argument ~a of type ~a (not ~a)" op (+ i 1) each-type actual-type)))]
-       [(? list? param-types)
-        (for/and ([arg exprs] [type param-types] [i (in-naturals)])
-          (define actual-type (expression->type arg env error!))
-          (unless (equal? actual-type type)
-            (error! stx "~a expects argument ~a of type ~a (not ~a)" op (+ i 1) type actual-type)))]
-       [_ (error "Operator has no type signature" op (length exprs))])
-     (get-rt-type op (length exprs))]
+     (define sigs (get-sigs op (length exprs)))
+     (unless sigs (error "Operator has no type signature" op (length exprs)))
```

Comment:
This will cause an error, since you're passing the arguments `op` and `(length exprs)` to a format string that expects no arguments.

Reply:
fixed format string